### PR TITLE
app-root, lib-user: Add search param validation for stats project_id, start_date, and end_date

### LINKED ISSUE #6168: Stats state url search param validation
There should be some validation/sanitation on the url search params for stats state - `project_id`, `start_date`, and `end_date`.

When a user enters a url for their user stats page or a group stats page with an invalid stats state search param, what should we do?

1. We could remove the param and show a simple browser alert noting something like "Invalid `project_id` included in the url has been removed"
2. We could remove the param without user notification
3. Other?

How much validation should we do?
- project_id
  - string
  - consisting of numbers
  - equal to or less than 5 characters? (currently I think project IDs are around 24000)
  - other?
- dates
  - string
  - YYYY-MM-DD format
  - start_date > end_date
  - other?

cc: @seanmiller26

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- app-root
- lib-user

## Linked Issue and/or Talk Post
- closes #6168 

## Describe your changes
- add project_id search param validation helper
- add end_date and start_date search params validation helper
- refactor app-root UserStats, Certificate, and GroupStats pages and containers
- refactor lib-user UserStats, Certificate, and GroupStats containers and components
- refactor MainContent with `paramsValidationMessage`
- refactor `selectedProject` or `projectId`, changing from 'AllProjects' as selected project ID when expect all projects, with `undefined` for when expect all projects

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX? n/a
- What user actions should my reviewer step through to review this PR?
  - app-root
    - review your user stats page (i.e. https://local.zooniverse.org:3000/users/markb-panoptes/stats)
      - review your user certificate page
    - review a group stats page (i.e. https://local.zooniverse.org:3000/groups/1326962)
- Which storybook stories should be reviewed?
  - lib-user stories:
    - Certificate, default: http://localhost:6008/?path=/story/components-certificate--default
    - Certificate, with invalid param: http://localhost:6008/?path=/story/components-certificate--params-validation-message
    - MainContent, default: http://localhost:6008/?path=/story/components-shared-maincontent--default
    - MainContent, no data: http://localhost:6008/?path=/story/components-shared-maincontent--no-stats
    - MainContent, with invalid param: http://localhost:6008/?path=/story/components-shared-maincontent--params-validation-message
    - Select: http://localhost:6008/?path=/story/components-shared-select--default
- Are there plans for follow up PR’s to further fix this bug or develop this feature? no

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

File: packages/app-root/src/utils/validateProjectIdParam.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  }
+  // project ID per search param should be a string
+  if (typeof projectId !== 'string') {
+    return {
+      projectId: null,
+      message: 'Invalid project_id, must be a string'
+    }
+  }
+  // as of August, 2024, production project IDs are ~24000 (5 digits)
+  // therefore, project IDs should be less than 5 digits
```
Comment (goplayoutside3) [inline packages/app-root/src/utils/validateProjectIdParam.js:14]:
```suggestion
  // therefore, project IDs should be less than 6 digits
```

File: packages/lib-user/src/components/shared/MainContent/MainContent.js [hunk 2]
```diff
@@ -231,11 +235,38 @@ function MainContent({
...
+              pad='medium'
+            >
+              <SpacedText uppercase={false}>No data found.</SpacedText>
+              <Text>
+                Start by{' '}
+                <Anchor href='https://www.zooniverse.org/projects'>
+                  classifying a project
+                </Anchor>
+                {' ' }now.
+              </Text>
```
Comment (goplayoutside3) [inline packages/lib-user/src/components/shared/MainContent/MainContent.js:261]: I think this `noStats` message should also include a suggestion like "Start by classifying a project now, or change the date range." It's likely volunteers visiting their stats page for the first time (and have not submitted a classification in the past 7 days) will see a 'no data' message before exploring the rest of the page. We should guide them to use the date range select.

**Review by goplayoutside3 — APPROVED** [review]
Looks good! I left a couple of non blocking comments below, but my primary suggestion would be to build a loading state for MainContent. I'll post a video below, but essentially now that there's a `noStats` state, when I'm viewing my stats page (or a group page) and I select different projects and date ranges, I see the `noStats` message while my stats are still loading. Even if the Loader component displays alongside the `noStats` message, that would be a more informative UX than no loading indicators.

https://github.com/user-attachments/assets/ac598d0a-35e0-46c9-a94e-761a20cd23a7

File: packages/lib-user/src/components/shared/MainContent/MainContent.js (hunk 2 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/shared/MainContent/MainContent.js:261]: Yes, great suggestion!

**Pull request merged**